### DISCUSSION TITLE
FastAPI SocketIO

### DISCUSSION THREAD

**Discussion by Forceres** [miguelgrinberg/python-socketio#1296] [General]
I've seen some kinds of SocketIO and FastAPI integrations, but all of them used a lot of global variables (objects). And I would like to know how can we avoid them or fully refuse to use them in that case?

**Comment by miguelgrinberg** [comment]
Not sure what integrations you have seen, but to use Socket.IO you need just a `Server` or `AsyncServer` instance. This is typically a global variable, but you can also embed it in a class, along with all the event handlers.

**Reply by Forceres** [reply]
Thanks for the answer, but can I embed it in a class, and then register events without global instance, for example, by class static method?

**Reply by miguelgrinberg** [reply]
Let me show you a couple of options.

Here is a standard mini-server, using a global `sio` variable:

```python
sio = socketio.AsyncServer()

@sio.event
async def connect(sid, environ, auth):
    print('incoming connection')
```

Here is the same using the factory function pattern:

```python
def create_socketio_server():
    sio = socketio.AsyncServer()

    @sio.event
    async def connect(sid, environ, auth):
        print('incoming connection')

    return sio
``` 

And here is an OOP version:

```python
class SocketIOServer:
    def __init__(self):
        self.sio = socketio.AsyncServer()
        self.sio.on('connect', self.on_connect)

    async def on_connect(self, sid, environ, auth):
        print('incoming connection')
```
